packages/yarn-plugin-allow-scripts/src/index.ts

```diff
@@ -0,0 +1,143 @@
+}
+
+// Yarn inserts spurious / for absolute paths on Windows
+const yarnCwdToPath = (cwd: string): string =>
+  // eslint-disable-next-line no-useless-escape
+  process.platform === 'win32' ? cwd.replace(/^[\/]/, '') : cwd
+
+const isPackageScriptAllowed = async (
+  project: ExtendedProject,
+  npm_package_json: string
+): Promise<boolean> => {
+  if (!project.allowScriptsConfig) {
+    project.allowScriptsConfig = new Map<string, boolean>()
+  }
+  const {
+    canonicalNamesByPath,
+    configs: { lifecycle },
+  } = await loadAllPackageConfigurations({
+    rootDir: yarnCwdToPath(project.cwd),
+  })
+  const scriptDir = dirname(npm_package_json)
+  const packageName = canonicalNamesByPath.get(scriptDir)
+  if (!packageName) {
+    process.stderr.write(
+      `FATAL ERROR: allow-scripts blocked execution of unexpected package script. ${JSON.stringify(
+        {
+          package: packageName,
+          manifestPath: npm_package_json,
+        }
+      )}`
+    )
+    process.exit(1)
+  }
+  if (lifecycle.disallowedPatterns.includes(packageName)) {
+    project.allowScriptsConfig.set(packageName, false)
+    return false
+  }
+  if (lifecycle.allowedPatterns.includes(packageName)) {
+    project.allowScriptsConfig.set(packageName, true)
+    return true
+  }
+  return null
+}
+
+const plugin: Plugin = {
+  hooks: {
+    wrapScriptExecution: (
+      executor: () => Promise<number>,
+      project: ExtendedProject,
+      _locator: Locator,
+      npm_lifecycle_event: string,
+      { env: { npm_package_json, npm_package_name } }: { env: any }
+    ): Promise<() => Promise<number>> => {
+      return new Promise((resolve, _reject) =>
+        isPackageScriptAllowed(project, npm_package_json).then((isAllowed) => {
+          if (!isAllowed) {
+            console.error(
+              isAllowed === null
+                ? `  allow-scripts blocking execution of unconfigured package script. ${JSON.stringify([npm_package_name, npm_lifecycle_event, npm_package_json])}`
+                : `  allow-scripts blocking execution of disallowed package script. ${JSON.stringify([npm_package_name, npm_lifecycle_event, npm_package_json])}`
+            )
+            return resolve(() => Promise.resolve(0))
+          }
+          return resolve(executor)
+        })
+      )
+    },
+    setupScriptEnvironment: (
+      project: ExtendedProject,
+      { npm_lifecycle_event, npm_package_json, npm_package_name }
+    ) => {
+      isPackageScriptAllowed(project, npm_package_json).then((isAllowed) => {
+        if (!isAllowed) {
+          console.error(
+            isAllowed === null
+              ? `  allow-scripts detected attempted execution of unconfigured package script. ${JSON.stringify([npm_package_name, npm_lifecycle_event, npm_package_json])}`
+              : `  allow-scripts detected attempted execution of disallowed package script. ${JSON.stringify([npm_package_name, npm_lifecycle_event, npm_package_json])}`
```

Comment:
> I'd avoid introducing the distinction here and print disallowed in both cases.
Missing configuration detection is happening elsewhere, and it's harder to miss there.

I'd like to keep the distinction - it's useful to immediately see why (missing or explicitly blocked).

> Also, the same check and message in two different hooks? Is this detecting different cases or will both be displayed for one package?

The one in `setupScriptExecutionEnvironment` should actually never happen since the script should have been blocked in the `wrapScriptExecution` hook. So this was as a warning that this was ineffective (hypothetical unforeseen bug in yarn or the plugin).


> Why is this hook not attempting to block anything? (It could process.exit if there's no other choice)

I think this makes sense. I have reworded the error messages and made the process exit here now.